src/coap_blockwise.c

```diff
@@ -63,65 +69,113 @@ static void on_block_sent(struct golioth_client *client,
-                                         on_block_sent,
-                                         ctx,
-                                         true,
-                                         GOLIOTH_SYS_WAIT_FOREVER);
+    enum golioth_status status = golioth_coap_client_set_block(client,
+                                                               ctx->path_prefix,
+                                                               ctx->path,
+                                                               ctx->is_last,
+                                                               ctx->content_type,
+                                                               ctx->offset,
+                                                               ctx->block_buffer,
+                                                               ctx->block_size,
+                                                               on_block_sent,
+                                                               ctx,
+                                                               true,
+                                                               GOLIOTH_SYS_WAIT_FOREVER);
+    ctx->status = status;
 }
 
 // Function to handle blockwise upload errors
-static enum golioth_status handle_upload_error()
+// This function retries for all errors like invalid state, failed mem alloc,
+// queue full, disconnects, etc returned by golioth_coap_client_set_block
+static enum golioth_status handle_upload_error(struct golioth_client *client,
+                                               struct blockwise_transfer *ctx)
 {
-    // TODO: handle errors like disconnects etc
-    return GOLIOTH_OK;
+    while (1)
+    {
+        GLTH_LOGE(TAG, "Blockwise upload failed with error %d. Retrying.", ctx->status);
+        golioth_sys_msleep(CONFIG_GOLIOTH_BLOCKWISE_RETRY_SLEEP_TIME_MS);
+        upload_single_block(client, ctx);
+        if (ctx->status == GOLIOTH_OK)
+        {
+            ctx->retry_count = 0;
+            break;
+        }
+        else
+        {
+            if (CONFIG_GOLIOTH_BLOCKWISE_RETRY_COUNT != -1)
+            {
+                ctx->retry_count++;
+                if (ctx->retry_count > CONFIG_GOLIOTH_BLOCKWISE_RETRY_COUNT)
+                {
+                    break;
+                }
+            }
+        }
+    }
+    return ctx->status;
 }
 
 // Function to manage blockwise upload and handle errors
 static enum golioth_status process_blockwise_uploads(struct golioth_client *client,
                                                      struct blockwise_transfer *ctx)
 {
-    enum golioth_status status = GOLIOTH_ERR_FAIL;
-    ctx->block_size = 0;
-    call_read_block_callback(ctx);
-    if (ctx->block_size)
+    enum golioth_status status = call_read_block_callback(ctx);
+    if (status == GOLIOTH_OK)
     {
-        status = upload_single_block(client, ctx);
+        upload_single_block(client, ctx);
         if (ctx->status == GOLIOTH_OK)
```

Comment:
It feels confusing to be mixing storing the status in the context and keeping it local on the stack. I'm not sure that we need it in the context struct.

Reply:
I see. The status in the context struct is used to store the status that we get from the server response that is passed to us via a callback executed on the CoAP client thread. But because we are doing synchronous operations, the server response status is also returned to us directly, so we don't even actually need the callback.

On the other hand, the synchronous functions internal to the coap client create and destroy a semaphore for each transaction (i.e. for each block that we download). It would be more efficient to use asynchronous transfers and to make the operations blocking by waiting on a semaphore that is managed by the blockwise transfers module and created once per blockwise transfer.